### PR TITLE
Support for test cases with multiple data input

### PR DESCRIPTION
**Pull request by MichaelHuth**
Should solve https://github.com/byte-physics/igor-unit-testing-framework/issues/5

This enabled multi data test cases.
It also works in combination with test cases that monitor background activity.

### CODE REVIEW

**Comment by MichaelHuth** [comment]
@t-b second revision with multi data test cases taking only a single argument of the same type as the wave result of associated data generator

**Comment by t-b** [comment]
@MichaelHuth FYI. Rebase against master to make CI pass. This is required due to gitlab runner retagging.

**Comment by t-b** [comment]
I've dived into the code and looks from the bird eye logic good!

Some things I noted:
- GetUTFTDGenerator should be a generic function for arbitrary tags. The matching should also be much stricter using a regex ala

```
•splitstring/E="//*[[:space:]]*\\QUTF_TD_DATAGENERATOR\\E(.*)$" str, str1; print V_flag; print str1
  1
   func
```

- Please use/create constants for the wave type return values
- Adding lists like `return testCaseList + testCaseMDList` fails if the first list is empty as in `";"`.
- Can you move the change `string testCaseList = FunctionList("!*_IGNORE", ";", "KIND:18,NPARAMS:0,VALTYPE:1,WIN:" + procWin)` into a separate commit and adapt the documentation? This is a compatibilty relevant change and I don't want to hide it.
- The dimension label reading from the data generator wave seems to be missing.
- Have you ideas for getting the testcase count for JUNIT/TAP correct?

**Comment by MichaelHuth** [comment]
> * Have you ideas for getting the testcase count for JUNIT/TAP correct?

For TAP I have to move the handling after the test cases into the do while loop.

For JUNIT, we have juProps.testCaseList that is at the moment only used to fill juProps.juTS.tests with the number of items in that list. So I can replace that with a testCaseCount field, that stores the number of test cases for this Suite.

To get the count I have to add a function to determine to complete number of test cases after

```
		testCaseList = getTestCasesMatch(procWin, testCase, enableRegExpTC, err)
		if(err & TC_LIST_EMPTY)
			continue
		endif
```
is called for each test suite. As the function had been verified before they went through the matching filter, I can rely on the DimSizes of the waves from the multi data test cases data generator functions.

**Comment by MichaelHuth** [comment]
@t-b The test case numbers for JUNIT and tap should be correct now. I dropped the = sign from the generator line and introduced a generic GetFunctionTag function. I also addressed the other comments.

**Comment by t-b** [comment]
`DisplayHelpTopic "Backslash in Regular Expressions"`

**Comment by MichaelHuth** [comment]
@t-b with the last commit also multi data support for reentry functions is implemented. The reentry function must have the same function format as the origin multi data test case or w/o multi data.
Should we also call incrError() if the UTF encounters a multi data reentry function of the wrong format?

**Comment by t-b** [comment]
@MichaelHuth Could you rebase against latest master so that CI passes?

**Comment by t-b** [comment]
Review:

General:
- Looks already really promising!
- [ ] example 12 and 13 are not linked in the documentation
- [ ] please change the module names of the new procedure files in internal_dev, they currently coincide with the examples.
- [ ] example 13 has mixed indentation
- [ ] How about cutting down example 13 to a minimum example? If you like to keep the current example I'm happy having it as example14 as well.

- [ ] Whitespace issues
```
$ git diff master.. --check
procedures/unit-testing-basics.ipf:2187: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2228: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2337: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2341: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2344: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2349: trailing whitespace.
+
```
- [ ] No need for semicolons
```
$ git grep ";$" *.ipf | grep -v "//"
unit-testing-basics.ipf:                SplitString/E=expr funcLine, funcTag;
```

File: docu/sphinx/source/advanced.rst [hunk 1]
```diff
@@ -250,6 +250,112 @@ the `_REENTRY` function.
...
+     SetDimLabel 0, 1, second, data
+     return data
+   End
+
+   Function testCase_REENTRY([var])
+	   variable var
+
+	   print "Reentered test case with argument ", var
+	   PASS()
+   End
```
Comment (t-b) [inline docu/sphinx/source/advanced.rst]: Indentation in the example looks odd.

**Comment by t-b** [comment]
> Should we also call incrError() if the UTF encounters a multi data reentry function of the wrong format?

Yes I think that this is a good idea. Same for wrong reentry format.

File: procedures/unit-testing-basics.ipf [hunk 2]
```diff
@@ -897,13 +972,122 @@ static Function TestCaseEnd(testCase, keepDataFolder)
...
+
+	numTC = ItemsInList(testCaseList)
+	for(i = 0; i < numTC; i += 1)
+		testCase = StringFromList(i, testCaseList)
+		FUNCREF TEST_CASE_PROTO TestCaseFunc = $testCase
+		if(UTF_FuncRefIsAssigned(FuncRefInfo(TestCaseFunc)))
+			tcCount += 1
+		else
+			dgenFuncName = GetFunctionTag(testCase, UTF_TD_GENERATOR, UTF_TD_GENERATOR_L)
+			dgenFuncName = GetFullFunctionName(err, dgenFuncName, procWin)
```
[the remark this answers is not in the document]
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: Yes, if the full function name can be retrieved is already checked before in CheckFunctionSignaturesTCMD(). Here it is just used to setup the globals before entering a multi data test case.

File: procedures/unit-testing-basics.ipf [hunk 3]
```diff
@@ -897,13 +972,122 @@ static Function TestCaseEnd(testCase, keepDataFolder)
...
+				printf "Could not get full function name of data generator: %s", dgen
+				continue
+			endif
+			FUNCREF TEST_CASE_PROTO_DGEN fDgen = $dgen
+			if(!UTF_FuncRefIsAssigned(FuncRefInfo(fDgen)))
+				printf "Data Generator function %s has wrong format. It is referenced by test case %s.\r", dgen, fullTestCaseMD
+				continue
+			endif
+			WAVE wGenerator = fDgen()
+			if(!((wType1 == 1 && WaveType(wGenerator, 1) == wType1 && WaveType(wGenerator) & wType0) || (wType1 != 1 && WaveType(wGenerator, 1) == wType1)))
```
[the remark this answers is not in the document]
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: Yes, missed that location when introducing the constants.

File: procedures/unit-testing-basics.ipf [hunk 4]
```diff
@@ -901,7 +901,7 @@ End
 static Function/S getTestCaseList(procWin)
 	string procWin
 
-	string fList = FunctionList("!*_IGNORE", ";", "KIND:18,NPARAMS:0,WIN:" + procWin)
+	string fList = FunctionList("!*_IGNORE", ";", "KIND:18,NPARAMS:0,VALTYPE:1,WIN:" + procWin)
```
[the remark this answers is not in the document]
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: Added that point

File: procedures/unit-testing-basics.ipf [hunk 5]
```diff
@@ -746,6 +799,28 @@ static Function CheckAbortCondition(abortCode)
 	endif
 End
 
+/// Check if function marked as Multi Data test case
```
[the remark this answers is not in the document]
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: Updated function comment

File: procedures/unit-testing-basics.ipf [hunk 6]
```diff
@@ -998,6 +1010,43 @@ static function GetTestCaseCount(testCaseList, procWin)
 	return tcCount
 End
 
+/// Checks functions signature of each multi data test case candidate
+/// returns 1 if ok, 0 otherwise
+/// when 1 is returned the wave type variable contain the format
+static Function GetFunctionSignatureTCMD(testCase, wType0, wType1)
```
[the remark this answers is not in the document]
Comment (MichaelHuth) [reply procedures/unit-testing-basics.ipf]: done

**Comment by t-b** [comment]
The indentation in the documentation is not correct:

grafik

And that code block is also missing in the output.

Please merge once that is resolved.

**Comment by t-b** [comment]
@MichaelHuth Don't merge. Example11-background.ipf is now broken. Please fix that, thanks.

**Comment by MichaelHuth** [comment]
I think this fixes it, it was introduced when converting the multi data decision logic to CallTestCase.

**Comment by t-b** [comment]
@MichaelHuth Thanks. I'll trust your testing.

**Pull request merged**